File: modules/eurobot2009.lua

-- Eurobot 2009: Atlantis

require('modules/colors')
module('eurobot2009', package.seeall)

-- Team colors
color1 = colors.ral_6018
color2 = colors.ral_3020

-- Match init
function init(fconf)
  -- Field configuration:
  --   columns: 1 to 10
  --   dispensers: (1 or 2) * 16
  -- Can be defined as 2-digit hexadecimal value: 0x<disp><col>
  local conf_col, conf_disp
  if fconf == nil or fconf < 0 then
    conf_col  = math.random(10)
    conf_disp = math.random(2)
  else
    conf_col  = fconf % 16 + 1
    conf_disp = math.floor(fconf/16) + 1
  end
  if conf_col < 1 or conf_col > 10 or conf_disp < 1 or conf_disp > 2 then
    error("invalid field configuration")
  end
  trace("Atlantis rules, columns: "..tostring(conf_col)..", dispensers: "..tostring(conf_disp))

  -- Various variables

  local table_size_x = 3.0
  local table_size_y = 2.1

  local col_space_x = 0.250
  local col_space_y = 0.200
  local col_offset_x = 0.400
  local col_offset_y = 0.125

  local disp_offset_x = 0.289
  local disp_offset_y = 0.250
  local disp_offset_z = 0.045

  local wall_width  = 0.022
  local wall_height = 0.070
  

  local o, ol, ols


  -- Ground
  trace("  ground")

  o = OGround(colors.ral_5015, color1, color2)
  o:add_to_world()

  -- Walls (N, E, W, small SE, small SW, plexi S)
  trace("  walls")

  o = OSimple()
  o:set_shape(Shape:box(table_size_x/2+wall_width, wall_width/2, wall_height/2))
  o:add_to_world()
  o:set_pos(0, table_size_y/2+wall_width/2, wall_height/2)
  o:set_color(colors.white)
  o = OSimple()
  o:set_shape(Shape:box(wall_width/2, table_size_y/2+wall_width, wall_height/2))
  o:add_to_world()
  o:set_pos(table_size_x/2+wall_width/2, 0, wall_height/2)
  o:set_color(colors.white)
  o = OSimple()
  o:set_shape(Shape:box(wall_width/2, table_size_y/2+wall_width, wall_height/2))
  o:add_to_world()
  o:set_pos(-table_size_x/2-wall_width/2, 0, wall_height/2)
  o:set_color(colors.white)

  o = OSimple()
  o:set_shape(Shape:box(wall_width/2, 0.100/2, wall_height/2))
  o:add_to_world()
  o:set_pos(0.900+wall_width/2, -table_size_y/2+0.050, wall_height/2)
  o:set_color(colors.white)
  o = OSimple()
  o:set_shape(Shape:box(wall_width/2, 0.100/2, wall_height/2))
  o:add_to_world()
  o:set_pos(-0.900-wall_width/2, -table_size_y/2+0.050, wall_height/2)
  o:set_color(colors.white)

  o = OSimple()
  o:set_shape(Shape:box(1.800/2+wall_width/2, config.draw_epsilon, 0.250/2))
  o:add_to_world()
  o:set_pos(0, -table_size_y/2, 0.125)
  o:set_color(colors.plexi)
  o = OSimple()
  o:set_shape(Shape:box(0.578/2+wall_width/2, config.draw_epsilon, wall_height/2))
  o:add_to_world()
  o:set_pos(1.200, -table_size_y/2, wall_height/2)
  o:set_color(colors.plexi)
  o = OSimple()
  o:set_shape(Shape:box(0.578/2+wall_width/2, config.draw_epsilon, wall_height/2))
  o:add_to_world()
  o:set_pos(-1.200, -table_size_y/2, wall_height/2)
  o:set_color(colors.plexi)


  -- Building areas
  trace("  building areas")

  o = OSimple()
  o:set_shape(Shape:box(1.800/2, 0.100/2, config.draw_epsilon))
  o:add_to_world()
  o:set_pos(0, 0.050-table_size_y/2, config.draw_epsilon)
  o:set_color(colors.ral_8017)

  o = OSimple()
  o:set_shape(Shape:box(0.600/2, 0.100/2, 0.030/2))
  o:add_to_world()
  o:set_pos(0, 0.050-table_size_y/2, 0.015)
  o:set_color(colors.ral_8017)

  o = OSimple()
  o:set_shape(Shape:cylinderZ(0.150, 0.060/2))
  o:add_to_world()
  o:set_pos(0, 0, 0.030)
  o:set_color(colors.ral_8017)


  -- Random column elements
  trace("  random columns elements")

  local col_placements = {
    {0,1,2}, {0,2,5}, {0,2,4}, {0,2,3}, {0,1,4},
    {0,1,5}, {0,4,5}, {0,1,3}, {0,3,5}, {0,3,4}
  }

  for i,j in ipairs(col_placements[conf_col]) do
    -- First team
    o = OColElem()
    o:add_to_world()
    o:set_color(color1)
    o:set_pos(-col_offset_x-(2-math.floor(j%3))*col_space_x, -col_offset_y+(3-math.floor(j/3))*col_space_y)
    o = OColElem()
    o:add_to_world()
    o:set_color(color1)
    o:set_pos(-col_offset_x-(2-math.floor(j%3))*col_space_x, -col_offset_y+math.floor(j/3)*col_space_y)
    -- Second team
    o = OColElem()
    o:add_to_world()
    o:set_color(color2)
    o:set_pos(col_offset_x+(2-math.floor(j%3))*col_space_x, -col_offset_y+(3-math.floor(j/3))*col_space_y)
    o = OColElem()
    o:add_to_world()
    o:set_color(color2)
    o:set_pos(col_offset_x+(2-math.floor(j%3))*col_space_x, -col_offset_y+math.floor(j/3)*col_space_y)
  end


  -- Dispensers
  trace("  dispensers")

  -- Fixed
  od = ODispenser()
  od:add_to_world()
  od:set_pos(table_size_x/2-disp_offset_x, -table_size_y/2, disp_offset_z, 2)
  for i = 1,5 do
    o = OColElem()
    o:add_to_world()
    o:set_color(color1)
    od:fill(o, i*0.035)
  end
  od = ODispenser()
  od:add_to_world()
  od:set_pos(disp_offset_x-table_size_x/2, -table_size_y/2, disp_offset_z, 2)
  for i = 1,5 do
    o = OColElem()
    o:add_to_world()
    o:set_color(color2)
    od:fill(o, i*0.035)
  end

  -- Random
  od = ODispenser()
  od:add_to_world()
  od:set_pos(table_size_x/2-wall_width/2, conf_disp==1 and disp_offset_y or -disp_offset_y, disp_offset_z, 1)
  for i = 1,5 do
    o = OColElem()
    o:add_to_world()
    o:set_color(color1)
    od:fill(o, i*0.035)
  end
  od = ODispenser()
  od:add_to_world()
  od:set_pos(-table_size_x/2+wall_width/2, conf_disp==1 and disp_offset_y or -disp_offset_y, disp_offset_z, 3)
  for i = 1,5 do
    o = OColElem()
    o:add_to_world()
    o:set_color(color2)
    od:fill(o, i*0.035)
  end


  -- Lintels and lintel storages
  trace("  lintels and lintel storages")

  ols = OLintelStorage()
  ols:add_to_world()
  ols:set_pos(-0.200, 0)
  ol = OLintel()
  ol:add_to_world()
  ol:set_color(color1)
  ols:fill(ol)

  ols = OLintelStorage()
  ols:add_to_world()
  ols:set_pos(-0.600, 0)
  ol = OLintel()
  ol:add_to_world()
  ol:set_color(color1)
  ols:fill(ol)

  ols = OLintelStorage()
  ols:add_to_world()
  ols:set_pos(0.200, 0)
  ol = OLintel()
  ol:add_to_world()
  ol:set_color(color2)
  ols:fill(ol)

  ols = OLintelStorage()
  ols:add_to_world()
  ols:set_pos(0.600, 0)
  ol = OLintel()
  ol:add_to_world()
  ol:set_color(color2)
  ols:fill(ol)
end

File: python/simulotter/eurobot2009.py

"""
Eurobot 2009: Temples of Atlantis.
"""

from _simulotter._eurobot2009 import *
import _simulotter as _so
import eurobot as _eb

from _simulotter import vec2 as _vec2, vec3 as _vec3
from eurobot import TABLE_SIZE, WALL_WIDTH, WALL_HEIGHT, beacon_pos


TEAM_COLORS = (_eb.RAL[6018], _eb.RAL[3020])

# field contants
COL_SPACE_XY = _vec2(0.250, 0.200)
COL_OFFSET_XY = _vec2(0.400, 0.125)
DISP_OFFSET = _vec3(0.289, 0.250, 0.045)

COL_POS = (
    (0,1,2), (0,2,5), (0,2,4), (0,2,3), (0,1,4),
    (0,1,5), (0,4,5), (0,1,3), (0,3,5), (0,3,4),
    )


class OGround(_so.OGround):
  def __init__(self):
    _so.OGround.__init__(self, _eb.RAL[5015], *TEAM_COLORS)


class Match(_eb.Match):
  """
  Gather match data.

  Attributes:
    ground -- OGround instance

  Field configuration
    col: 0 to 9
    disp: 0 or 1

  """

  class Conf:
    def __init__(self, col, disp):
      assert 0 <= col <= 9
      assert disp in (0,1)
      self.col, self.disp = col, disp

    @classmethod
    def random(cls):
      import random
      return cls(random.randint(0,9), random.randint(0,1))

  def prepare(self, fconf=None):
    if fconf is None:
      self.conf = self.Conf.random()
    else:
      self.conf = self.Conf(*fconf)

    ph = self.physics

    # Ground and raised zone
    ground = OGround()
    ground.addToWorld(ph)
    self.ground = ground

    # Walls (N, E, W, S, small SE, small SW, plexi S)
    sh = _so.ShBox(_vec3(TABLE_SIZE.x+2*WALL_WIDTH, WALL_WIDTH, WALL_HEIGHT)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(0, TABLE_SIZE.y+WALL_WIDTH, WALL_HEIGHT)/2
    o.color = _so.Color.white
    sh = _so.ShBox(_vec3(WALL_WIDTH, TABLE_SIZE.y+2*WALL_WIDTH, WALL_HEIGHT)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(TABLE_SIZE.x+WALL_WIDTH, 0, WALL_HEIGHT)/2
    o.color = _so.Color.white
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(-TABLE_SIZE.x-WALL_WIDTH, 0, WALL_HEIGHT)/2
    o.color = _so.Color.white

    sh = _so.ShBox(_vec3(WALL_WIDTH, 0.100, WALL_HEIGHT)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(1.800+WALL_WIDTH, -TABLE_SIZE.y+0.100, WALL_HEIGHT)/2
    o.color = _so.Color.white
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(-1.800-WALL_WIDTH, -TABLE_SIZE.y+0.100, WALL_HEIGHT)/2
    o.color = _so.Color.white

    sh = _so.ShBox(_vec3(1.800+WALL_WIDTH, ph.margin_epsilon, 0.250)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(0, -TABLE_SIZE.y/2, 0.125)
    o.color = _so.Color.plexi
    sh = _so.ShBox(_vec3(0.578+WALL_WIDTH, ph.margin_epsilon, WALL_HEIGHT)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(2.400, -TABLE_SIZE.y, WALL_HEIGHT)/2
    o.color = _so.Color.plexi
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(-2.400, -TABLE_SIZE.y, WALL_HEIGHT)/2
    o.color = _so.Color.plexi


    # Building areas
    sh = _so.ShBox(_vec3(1.800, 0.100, ph.margin_epsilon)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(0, 0.050-TABLE_SIZE.y/2, ph.margin_epsilon)
    o.color = _eb.RAL[8017]

    sh = _so.ShBox(_vec3(0.600, 0.100, 0.030)/2)
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(0, 0.050-TABLE_SIZE.y/2, 0.015)
    o.color = _eb.RAL[8017]

    sh = _so.ShCylinderZ(_vec3(0.150, 0.150, 0.060/2))
    o = _so.OSimple(sh)
    o.addToWorld(ph)
    o.pos = _vec3(0, 0, 0.030)
    o.color = _eb.RAL[8017]


    # Random column elements
    for j in COL_POS[self.conf.col]:
      # first team
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[0]
      o.pos = COL_SPACE_XY * _vec2(j%3-2, 3-j//3) - COL_OFFSET_XY
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[0]
      o.pos = COL_SPACE_XY * _vec2(j%3-2,   j//3) - COL_OFFSET_XY
      # second team
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[1]
      o.pos = (COL_SPACE_XY * _vec2(j%3-2, 3-j//3) - COL_OFFSET_XY)*_vec2(-1,1)
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[1]
      o.pos = (COL_SPACE_XY * _vec2(j%3-2,   j//3) - COL_OFFSET_XY)*_vec2(-1,1)


    # Dispensers

    # Fixed
    od = ODispenser()
    od.addToWorld(ph)
    od.setPos( DISP_OFFSET*_vec3(-1,0,1) - TABLE_SIZE/2*_vec2(-1,1), 2 )
    for i in range(4):
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[0]
      od.fill(o, (i+1)*0.035)
    od = ODispenser()
    od.addToWorld(ph)
    od.setPos( DISP_OFFSET*_vec3(1,0,1) - TABLE_SIZE/2, 2 )
    for i in range(4):
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[1]
      od.fill(o, (i+1)*0.035)

    # Random
    ky = 1 if self.conf.disp == 0 else -1
    od = ODispenser()
    od.addToWorld(ph)
    od.setPos( _vec3(TABLE_SIZE.x-WALL_WIDTH, DISP_OFFSET.y, DISP_OFFSET.z)*_vec3(.5,ky,1), 1 )
    for i in range(4):
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[0]
      od.fill(o, (i+1)*0.035)
    od = ODispenser()
    od.addToWorld(ph)
    od.setPos( _vec3(TABLE_SIZE.x-WALL_WIDTH, DISP_OFFSET.y, DISP_OFFSET.z)*_vec3(-.5,ky,1), 3 )
    for i in range(4):
      o = OColElem()
      o.addToWorld(ph)
      o.color = TEAM_COLORS[1]
      od.fill(o, (i+1)*0.035)


    # Lintels and lintel storages
    ols = OLintelStorage()
    ols.addToWorld(ph)
    ols.setPos(-0.200, 0)
    ol = OLintel()
    ol.addToWorld(ph)
    ol.color = TEAM_COLORS[0]
    ols.fill(ol)

    ols = OLintelStorage()
    ols.addToWorld(ph)
    ols.setPos(-0.600, 0)
    ol = OLintel()
    ol.addToWorld(ph)
    ol.color = TEAM_COLORS[0]
    ols.fill(ol)

    ols = OLintelStorage()
    ols.addToWorld(ph)
    ols.setPos(0.200, 0)
    ol = OLintel()
    ol.addToWorld(ph)
    ol.color = TEAM_COLORS[1]
    ols.fill(ol)

    ols = OLintelStorage()
    ols.addToWorld(ph)
    ols.setPos(0.600, 0)
    ol = OLintel()
    ol.addToWorld(ph)
    ol.color = TEAM_COLORS[1]
    ols.fill(ol)


